### DISCUSSION TITLE
Retrieving keys for old backups stored on computer

### DISCUSSION THREAD

**Discussion by ghost** [ElDavoo/wa-crypt-tools#119] [General]
**Problem 1**:
I don't have the key files to all of my backups. But since they always worked when I needed to restore them after a phone reset or change, I thought I could simply restore them at _any given point_. I was now shocked to learn that my current WhatsApp version only restores my December 2023 backup (the one I had made before I messed up all my chats by switching OS systems). My February 2022 backup is not even found anymore, even though I copied all the files in the correct places, let alone older backups such as 2018, 2016 or 2015.

**Conclusion drawn:**
The time span to be able to restore a backup with local database files must be relatively short (probably 3-5 months).
 
**Question 1:**
How did WhatsApp get the keys for my backup if I never physically backuped the keys and yet the backups could always be restored shortly after?? I assume that they use the combination "backup files + phone number" and then assign/generate a new key to me? Or instead of generating a NEW key, do they store the OLD key on their servers and just download them to my phone whenever they see the combination of "correct phone number + correct backup files"?

This idea led me to the following hypothesis:
"_Maybe my key is unique and universal, so it probably works for ALL my backups that have the combination "backup files + correct phone number_". 

**Conclusion drawn:**
Yes, to some degree. While the official WA version doesn't want to recognize backups that are older than 3-5 months, the key that I extracted from the December 2023 backup (right before I messed up my chats) WORKED for my 2022 backups too, which the official WA version didn't even detect. But then the key stopped working some time around backup April 2022 and further back.

**Question 2:**
What's behind this? Why can I decrypt backups with my Dec. 2023 backup key up to more than 1.5 years further back but then this key suddenly stops working for the other backups?

**Question 3**:
The only way to get a key for all of my backups (as I have never copied the key files) is to get WhatsApp to load and restore them and then a new key will be assigned OR the old key will be retrieved from their servers, right? Or is there any other way to get access to my backups? Unfortunately, WhatsApp doesn't detect backups older than a couple months.

**Question 4**
concerning the 64-letters E2EE-Backup-Encryption key: is this key truly a complete replacement of the former local key-file or is it just an ADDITIONAL protection to the local key file? Like: if I now enable E2EE backup protection and generate a 64-key - will this key from now on until eternity serve as the MASTER KEY for all my future backups or will I **still need** the local key-file in addition? The way I understood it is that it's only an additional protection to the local key and there's no external decryption possible with ONLY the 64-key and without the original key file?

Is there a way to get WhatsApp to restore my old backups so that I can get a new key for them or are they lost for good? I'm still not sure how the process works - I think WA stores it on its servers or else backup restoring without the local key files (which I never had) wouldn't have been possible. So they probably also have all those key files for my older backups on their server? But then why doesn't WA detect them?

The [offical WA customer service webpage](https://faq.whatsapp.com/6181521285295518/?cms_platform=android) says that in order to restore local backup files you have to be using Android 9.0 or below. I've tried it with a phone running on Android 8.0, but WA still wouldn't detect any backup files older than Dec. 2023...

Thanks a lot guys!!

**Comment by ElDavoo** [comment]
Hello, 
> The time span to be able to restore a backup with local database files must be relatively short (probably 3-5 months).  

Or maybe, every time something like a restore happens, a new key gets generated.  
This is why it's better to generate your E2E key.

> December 2023 backup (right before I messed up my chats) WORKED for my 2022 backups too, which the official WA version didn't even detect.
> But then why doesn't WA detect them?  

Something makes your backup invalid, remember that WA does not only need to decrypt the backup but also to open it, check it, migrate the schema... Maybe that's not possible to do if also something is stored in the server (?). You can check the difference between your backups with `wainfo`, maybe you can bisect and find something different in the header.

With `waencrypt` you can technically wrap old DB files in new headers. What would happen next is probably the app will tell you your backups is damaged but you might be lucky.


> I now enable E2EE backup protection and generate a 64-key - will this key from now on until eternity serve as the MASTER KEY for all my future backups
Yes, I really advise to do that if you want to self-manage your backups.

**Comment by ElDavoo** [comment]
> So many questions...  

Yet no answer, this project is about doing things once you have the key, I don't know anything about WA's key shenanigans.  

> That said, I'm rather new to all of this coding stuff, so I may probably need some friends of mine to help who know a lot more about computer science and coding.

You have curiosity, that's all that's needed. Now you need practice. 

> HOWEVER: after restoring is "finished" at 100%, I get a message saying the backup could not be restored, 0 messages restored.

And that's what's probably going to happen if you do the re-encrypting operation i described earlier

**Comment by ElDavoo** [comment]
Anyway, the topic of merging databases is discussed [here](https://github.com/ElDavoo/wa-crypt-tools/discussions/118), you might want to take a read.
